rust_icu_ecma402/src/listformat.rs

```diff
@@ -0,0 +1,234 @@
+// limitations under the License.
+
+//! Implements the traits found in [ecma402_traits::listformat].
+
+use ecma402_traits;
+use ecma402_traits::listformat;
+use rust_icu_common as common;
+use rust_icu_ulistformatter as ulfmt;
+use std::fmt;
+
+/// Implements [listformat::Format] using ICU as the underlying implementing library.
+#[derive(Debug)]
+pub struct Format {
+    rep: ulfmt::UListFormatter,
+}
+
+// Full support for styled formatting is available since v67.
+#[cfg(feature = "icu_version_67_plus")]
+pub(crate) mod internal {
+    use rust_icu_sys as usys;
+    use ecma402_traits::listformat::options;
+
+    // Converts the param types from ECMA-402 into ICU.
+    pub fn to_icu_width(style: &options::Style) -> usys::UListFormatterWidth {
+        use options::Style;
+        use usys::UListFormatterWidth;
+        match style {
+            Style::Long => UListFormatterWidth::ULISTFMT_WIDTH_WIDE,
+            Style::Short => UListFormatterWidth::ULISTFMT_WIDTH_SHORT,
+            Style::Narrow => UListFormatterWidth::ULISTFMT_WIDTH_NARROW,
+        }
+    }
+
+    // Converts the Type param from ECMA-402 into ICU.
+    pub fn to_icu_type(in_type: &options::Type) -> usys::UListFormatterType {
+        use options::Type;
+        use usys::UListFormatterType;
+        match in_type {
+            Type::Conjunction => UListFormatterType::ULISTFMT_TYPE_AND,
+            Type::Disjunction => UListFormatterType::ULISTFMT_TYPE_OR,
+        }
+    }
+}
+
+impl listformat::Format for Format {
+    type Error = common::Error;
+
+    /// Creates a new [Format], from a [ecma402_traits::Locale] and [listformat::Options].
+    fn try_new<L: ecma402_traits::Locale>(
+        l: L,
+        _opts: listformat::Options,
+    ) -> Result<Format, Self::Error> {
+        let locale = format!("{}", l);
+
+        #[cfg(feature = "icu_version_67_plus")]
+        {
+            let width = internal::to_icu_width(&_opts.style);
+            let in_type = internal::to_icu_type(&_opts.in_type);
+            let rep = ulfmt::UListFormatter::try_new_styled(&locale, in_type, width)?;
+            Ok(Format { rep })
+        }
+
+        // The non-v67 implementation is less featureful.
+        #[cfg(not(feature = "icu_version_67_plus"))]
+        {
+            let rep = ulfmt::UListFormatter::try_new(&locale)?;
+            Ok(Format { rep })
+        }
+    }
+
+    /// Formats the given string.
+    fn format<I, L, W>(self, list: L, f: &mut W) -> fmt::Result
+    where
+        I: fmt::Display,
+        L: IntoIterator<Item = I>,
+        W: fmt::Write,
+    {
+        // This is an extremely naive implementation: collects the list into
+        // a slice and formats that thing.
+        let list_str: Vec<String> = list.into_iter().map(|e| format!("{}", e)).collect();
+        let refs: Vec<&str> = list_str.iter().map(|e| e.as_str()).collect();
+        let result = self.rep.format(&refs[..]).map_err(|e| e.into())?;
+        write!(f, "{}", result)
```

Comment:
Ah, that's a bit sad, but I guess that's the cost of FFI here. It's ok, it shouldn't matter that much for now.

One thing you could do is use `e.to_string()` instead of `format!`, but that's minor.

Reply:
There's good news and bad news here.

- Good news is that there is an ICU API which is friendlier to the iterated approach than this one.  I didn't expose it because it was harder to use in general even though in this particular spot it's a better fit.  Either way we can fix that at any point and the user's won't see the difference.
- Bad news is, your remark still stands, the two APIs aren't the best match.  Given this code is a polyfill for icu4x, it may be fine to let smaller mismatches go.  

But my impression is that it should be fine for now.